### ISSUE TITLE
E2E testing using protractor

### ISSUE THREAD

**Issue by PaulL1** [ngbp/ngbp#196]
As noted in #173, the AngularJS e2e documentation (http://docs.angularjs.org/guide/dev_guide.e2e-testing) says that protractor is the new e2e testing framework.  

I'm working towards a pull request that merges in protractor as the e2e testing framework.  I'll try to step through this slowly, I'm starting with documenting the process I'm using to install and use protractor, which should give people a chance to comment on it if they wish, I'll finish off with a pull request that installs protractor, selenium and chromedriver into the current NGBP setup, and provides a couple of sample tests against the sample app.  These will need to be ported across to the new boilerplate 4, but I'm presuming having it at least working is a step towards that.

The first step is acquiring everything you need, which means getting protractor, a selenium-standalone jar file, and a chromedriver jar file.  The simplest way I've found so far is:

```
sudo npm install -g protractor
/usr/local/lib/node_modules/protractor/bin/install_selenium_standalone
```

The last command uses a protractor script to install selenium and chromedriver into the ./selenium directory, which isn't ideal but is what the tool does.  (Obviously the command location will depend on your OS, I don't see a way to generalise it as yet).

Then create a configuration file.  There are a lot of options, I'm choosing to have protractor autostart the selenium server for simplicity, even though this is a bit slower.  I think end to end tests won't be run as frequently as unit tests, so the startup time is a reasonable tradeoff for simplicity.  I'm following the karma pattern, so I'm putting this into:

```
protractor/protractor-scenario.tpl.js
```

```
exports.config = {
  // ----- How to setup Selenium -----
  //
  // There are three ways to specify how to use Selenium. Specify one of the
  // following:
  //
  // 1. seleniumServerJar - to start Selenium Standalone locally.
  // 2. seleniumAddress - to connect to a Selenium server which is already
  //    running.
  // 3. sauceUser/sauceKey - to use remote Selenium servers via SauceLabs.

  // The location of the selenium standalone server .jar file.
  seleniumServerJar: './selenium/selenium-server-standalone-2.37.0.jar',
  // The port to start the selenium server on, or null if the server should
  // find its own unused port.
  seleniumPort: null,
  // Chromedriver location is used to help the selenium standalone server
  // find chromedriver. This will be passed to the selenium jar as
  // the system property webdriver.chrome.driver. If null, selenium will
  // attempt to find chromedriver using PATH.
  chromeDriver: './selenium/chromedriver',
  // Additional command line options to pass to selenium. For example,
  // if you need to change the browser timeout, use
  // seleniumArgs: ['-browserTimeout=60'],
  seleniumArgs: [],

  // If sauceUser and sauceKey are specified, seleniumServerJar will be ignored.
  // The tests will be run remotely using SauceLabs.
  sauceUser: null,
  sauceKey: null,

  // The address of a running selenium server. If specified, Protractor will
  // connect to an already running instance of selenium. This usually looks like
  // seleniumAddress: 'http://localhost:4444/wd/hub'
  seleniumAddress: null,

  // The timeout for each script run on the browser. This should be longer
  // than the maximum time your application needs to stabilize between tasks.
  allScriptsTimeout: 11000,

  // ----- What tests to run -----
  //
  // Spec patterns are relative to the location of this config.
  specs: [
    '../src/**/*.scenario.js',
  ],

  // ----- Capabilities to be passed to the webdriver instance ----
  //
  // For a full list of available capabilities, see
  // https://code.google.com/p/selenium/wiki/DesiredCapabilities
  // and
  // https://code.google.com/p/selenium/source/browse/javascript/webdriver/capabilities.js
  capabilities: {
    'browserName': 'chrome'
  },

  // ----- More information for your tests ----
  //
  // A base URL for your application under test. Calls to protractor.get()
  // with relative paths will be prepended with this.
  baseUrl: 'http://localhost:3000',

  // Selector for the element housing the angular app - this defaults to
  // body, but is necessary if ng-app is on a descendant of <body>  
  rootElement: 'body',

  // A callback function called once protractor is ready and available, and
  // before the specs are executed
  // You can specify a file containing code to run by setting onPrepare to
  // the filename string.
  onPrepare: function() {
    // At this point, global 'protractor' object will be set up, and jasmine
    // will be available. For example, you can add a Jasmine reporter with:
    //     jasmine.getEnv().addReporter(new jasmine.JUnitXmlReporter(
    //         'outputdir/', true, true));
  },

  // The params object will be passed directly to the protractor instance,
  // and can be accessed from your test. It is an arbitrary object and can
  // contain anything you my need in your test.
  // This can be changed via the command line as:
  //   --params.login.user 'Joe'
  params: {
    login: {
      user: 'Jane',
      password: '1234'
    }
  },

  // ----- Options to be passed to minijasminenode -----
  //
  // See the full list at https://github.com/juliemr/minijasminenode
  jasmineNodeOpts: {
    // onComplete will be called just before the driver quits.
    onComplete: null,
    // If true, display spec names.
    isVerbose: false,
    // If true, print colors to the terminal.
    showColors: true,
    // If true, include stack traces in failures.
    includeStackTrace: true,
    // Default time to wait in ms before a test fails.
    defaultTimeoutInterval: 30000
  }
};
```

You can then run protractor with:

```
protractor protractor/protractor-scenario.tpl.js
```

You need a scenario before it does anything, so I'm creating a file in: 

```
home/home.scenario.js
```

```
/**
 * End to end tests for the home page
 */

var HomePage = function() {
  this.home = element(by.id('home'));
  this.about = element(by.id('about'));
  this.readme = element(by.id('readme'));
  this.gitHub = element(by.id('gitHub'));
  this.support = element(by.id('support'));

  this.get = function() {
    browser.get('index.html');
  };  
};

describe( 'Navigate to home page', function() {
  it ( 'should allow navigation to the home page', function() {
    var homePage = new HomePage();

    homePage.get();
    expect(homePage.home.getText()).toEqual('Home');
    expect(homePage.about.getText()).toEqual('What is it?');
    expect(homePage.readme.getText()).toEqual('Read the Docs');
    expect(homePage.gitHub.getText()).toEqual('Github');
    expect(homePage.support.getText()).toEqual('Support');
    expect(browser.getTitle()).toEqual('Home | ngBoilerplate');

    homePage.about.click();

    expect(browser.getTitle()).toEqual('What is It? | ngBoilerplate');    
  }); 
});
```

We also need to edit .gitignore, and ignore the selenium directory, we don't want those files in git (they're large, and don't belong there):

```
+ selenium/
```

**Comment by PaulL1** [comment]
So the commit above works (within limits.....).

I've verified through cloning the repository, then:

``` sh
$ npm install
$ bower install
$ /usr/local/lib/node_modules/protractor/bin/install_selenium_standalone
$ grunt build
```

The grunt build will throw an error, I'll talk about that in a second, but it builds.  I then link the build directory across to my web server (localhost:3000), and run:

``` sh
$ protractor protractor/protractor-scenario.tpl.js
```

It seems to run happily.

There are four issues that I have that I think need resolving before this could be merged:
1. Karma wants to run the scenario files.  I can't work out why, but this means that karma is failing when running grunt.  I've poked around without finding where I change this, but I suspect someone else will know exactly why this happens and how to change it
2. I've chosen to use the page objects format as described on the protractor getting started page: https://github.com/angular/protractor/blob/master/docs/getting-started.md.  This really doesn't work well without the ability to include one javascript file in another.  Google tells me this is hard, but I have an inkling that some of the AngularJS module functionality will let me do this.  I just don't have my head straight on how to access that in a test script, rather than in "real" angular code.  It would be a better sample if I had that sorted
3. I don't really like the selenium install process.  There must be a better way, but I can't see it as yet.  Perhaps when it gets integrated into Angular proper it'll come with a better installer.  Really it needs an npm task for selenium and chromedriver
4. I need to either find a way to convince protractor that it'd like to run directly against the build directory (rather than against a web server), or provide some better instruction to the casual user as to how they get this into their web server.  Although arguably this is something they would need to know already anyway.

Anyway, at this point, any help or comments gratefully received.

**Comment by PaulL1** [comment]
I've further updated the pull request having sorted the karma/scenario files issue.  There is still some work to do here, but in summary I have:
1. In karma/karma-unit.tpl.js, excluded the scenario files from those picked up and loaded into the browser
2. In build.config.js, created new jse2e and coffeee2e file lists, and excluded the scenario files from the app files
3. In Gruntfile.js, added the jse2e files to the js linter, and the coffeee2e files to the coffee linter

This now allows protractor to run without breaking grunt.  However, the clear outstanding is that I need a grunt task for running protractor.  That is on my list of things to look at, albeit that I expect whatever I imagine to change as the Angular folks integrate protractor more closely.

**Comment by PaulL1** [comment]
And a further update.  I keep thinking I need to go do some housework before the upcoming dinner party.  But I'm doing OK at ignoring it.  

So the grunt integration comes with grunt-protractor-runner.  Install with

```
npm install grunt-protractor-runner --save-dev
```

I've updated the commit to include this in the package.json, to add extra sections to the gruntfile (creating a new task "grunt e2e") and to fix a couple of defects in the previous commit I submitted (typos in the gruntfile).

This seems to work well enough.  It'd be nice to have a watch style command for the e2e tests, but I'll not get ahead of myself there, it clearly shouldn't go in the normal watch as that'd be pretty slow for everyone.

**Comment by tony-kerz** [comment]
paul, thank you so much for contributing this work!

i'm really new to angular and it's node based eco-system, but i really like the ngbp value proposition especially for peeps like myself trying to find quality workability patterns for the stack.

to minimize the volume of elements to study, my strategy is to focus on those currently popular, and i was just reading about the angular team migrating to protractor for testing, so i figured i'd focus my studies there, but i was just lamenting the thought of attempting to modify ngbp myself for protractor integration, so your contribution is timely.

the least (and perhaps the most) i can do to assist is clone out your solution and give it a go from a test user perspective. i'll let you know how it goes and thanks again!

**Comment by PaulL1** [comment]
On the matter of modularising the tests somewhat, I've found astrolab, which is an additional plugin for protractor, and looks like it provides a require syntax to share content between tests.  https://github.com/stuplum/astrolabe.  It's on my list to try to integrate into this and the examples over the next couple of days.  I've got it working in my related tutorial that I use for proof of concept (http://technpol.wordpress.com/2013/11/16/5-end-to-end-testing/), I'm next going to update my pull request to include this.

**Comment by PaulL1** [comment]
OK, so we can install astrolabe like this:

``` sh
$ npm install --save-dev astrolabe
```

We then modify our scenario files to split out into page objects and scenarios.  The page objects are reusable, and can have substantial helper methods on them (things like "login" or "register"), allowing us to write quite high level scripts.  The ngbp demo app doesn't really give a lot of opportunity to demonstrate this, but I've broken it out anyway to provide an illustration of how it works.

I've chosen to call the page object files xxx.part.scenario.js, with the scenarios themselves just xxx.scenario.js.  I'm not sure whether that naming convention meets with everyone's pleasure, but it's not really baked into the code anywhere, it's just a convention and people could pick different ones.

The pull request is updated to provide updates on the readme file, and the additional scenario files.

The key outstanding item is some clear way to tell people that they need to update the protractor configuration before it will run - they need to put their url into the configuration.  Happy testing!!

**Comment by tony-kerz** [comment]
i cloned the repo prior to astrolabe and just kicked the tires on the protractor functionality which seemed to work fine.

one thing i don't quite get is the workflow w.r.t. the server running at baseUrl,
seems like that currently has to be manually started...?

ideally, in order to automate the tests using a continuous-integration server (e.g. [jenkins](jenkins-ci.org)),
we would want an instance of that server spun up for the test and then spun down when finished.

i noticed your server looks like a rails app (uses port 3000), so it would probably be more involved to spin that up and down, but in many cases single-page-apps can consist solely of static content which i think can be served up (and spun down) readily via [grunt-contrib-connnect](https://github.com/gruntjs/grunt-contrib-connect).

so maybe, an ngbp which includes e2e functionality should contain integration with grunt-contrib-connect...?

**Comment by PaulL1** [comment]
Tony - might make sense.  You're right that I'm using rails, and for me I have guard running in the background, so the rails server is always there.  I've actually added grunt-shell to my project, and I use that to call rake db:test:prepare to reset my database before running the e2e tests, which again is something that people would probably need for whatever their server is.  I also run a standalone selenium rather than spinning it up every time, which makes it much more convenient.

I agree that perhaps we just need a lightweight web server that is always there so as to have a solution that is running "out of the box", and therefore we would use the baseUrl that goes with that server.  People could then tailor from there depending on what sort of backend they were using.

I'm making good progress with astrolabe, and I really like it.  It really was the bit that was necessary to make it sensible.

**Comment by PaulL1** [comment]
A further note, and I need to submit a pull request with a documentation update to reflect.  E2e testing runs quite a bit slower than unit testing, so subsetting the test cases is useful.  You can also run a single spec through (for example):

``` sh
grunt e2e --specs=src/app/home/home.scenario.js
```

**Comment by joshdmiller** [comment]
@PaulL1 - We're hard at work on v0.4, for which Protractor will run the E2E tests. There are a lot of features left to implement (including the E2E testing) over the next two or three weeks until the final release and I'd love some additional help if you've got some free time. If you're interested in helping out, shoot me an email (click my name and you should see my email address if you're logged in).

The work is all on separate branches and repos. I won't go into too much detail here (for the sake of brevity), but check out the [v0.4-wip branch](https://github.com/ngbp/ng-boilerplate/tree/v0.4-wip) and the various accompanying [ngbp repositories](https://github.com/ngbp) like [ngbp-contrib-less](https://github.com/ngbp/ngbp-contrib-less) to get a sense of how radically, insanely different the new version is.

Psst - I'm whispering here, but the super-unstable alphas are also in NPM. Try `npm install -g ngbp-cli` and then in an empty directory try `ngbp init` to get a new empty project (probably won't work in Windows - I don't know).

**Comment by aruizca** [comment]
+1

**Comment by revolunet** [comment]
+1

**Comment by fraserxu** [comment]
+1

**Comment by brijeshsoni** [comment]
I have an application in angularJs and in that I have 1 test case in protractor framework.I want to call a backend action(Java) from protractor test case.How can I do that.
ajax, http, jquery actions are not working
Please help with an example

**Comment by PaulL1** [comment]
@brijeshsoni: I'm not sure that's an ngbp-specific question, it sounds like a generic protractor question.  In my experience you'll have some limitations, protractor is mostly a UI driven testing tool.  So if you can't do it through the UI (i.e. by auto-clicking in the browser), then you probably can't do it.  I know there are some workarounds, like building wrapper pages that you can then use to hit the backend, or if your web services are restful you may be able to exercise them by going directly to that url in your browser (for get type services).  But in general the aim of protractor is to exercise the application as an end-user would, so it doesn't provide a lot of options for hitting backend services directly.

**Comment by ErichBSchulz** [comment]
Hi - for those impatient for e2e testing - this seems to be working ok:

https://github.com/ErichBSchulz/ngbp/commit/af3556088c2a07b4c3210ec99588c315a703629a

**Comment by ErichBSchulz** [comment]
@PaulL1 - re this:

> I've chosen to call the page object files xxx.part.scenario.js, with the scenarios themselves just xxx.scenario.js. 

how about `xxx.page.scenario.js`? too obvious?

**Comment by PaulL1** [comment]
Reasonable. Ngboilerplate doesn't need to know about parts anyway - it just thinks they're more scenarios, so it only impacts sample code and doco.

**Comment by taufansatrio** [comment]
How is the v0.4 going? I've seen very little activity recently
If I need to add e2e support for the current version, what is the most recommended (and preferably simplest) way to do so? Thanks
